### PR TITLE
222 job results details

### PR DESCRIPTION
**Pull request by keithjgrant**
Addresses #4223

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

### CODE REVIEW

File: awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx [hunk 1]
```diff
@@ -1,35 +1,156 @@
...
-            component={Link}
-            to="/jobs"
-          >
-            {i18n._(t`Close`)}
-          </Button>
-        </ActionButtonWrapper>
-      </CardBody>
-    );
-  }
+function JobDetail({ job, i18n }) {
```
Comment (keithjgrant) [inline awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx:32]: I went back and forth a lot about whether to make a different component for each type of job. In the end it just felt like way too much repeated code with very little variation. We can always split them back out down the line if the different types end up deviating further from each other.

File: awx/ui_next/src/types.js [hunk 2]
```diff
@@ -65,8 +65,61 @@ export const QSConfig = shape({
...
+});
+
+export const Credential = shape({
+  id: number.isRequired,
+  name: string.isRequired,
+  cloud: bool,
+  kind: string,
+});
+
+export const Job = shape({
```
Comment (jakemcdermott) [inline awx/ui_next/src/types.js:101]: nod

File: awx/ui_next/src/screens/Job/JobDetail/JobDetail.test.jsx [hunk 3]
```diff
@@ -1,22 +1,60 @@
...
+
+    const wrapper = mountWithContexts(<JobDetail job={job} />);
+    const details = wrapper.find('Detail');
+
+    function assertDetail(detail, label, value) {
+      expect(detail.prop('label')).toEqual(label);
+      expect(detail.prop('value')).toEqual(value);
+    }
+
+    assertDetail(details.at(0), 'Status', 'Successful');
```
Comment (marshmalien) [inline awx/ui_next/src/screens/Job/JobDetail/JobDetail.test.jsx:39]: We could make these tests more robust by checking if the correct string is rendered in the DOM versus checking an implementation detail (props).

File: awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.test.jsx [hunk 4]
```diff
@@ -113,19 +113,21 @@ describe('<JobTemplateDetail />', () => {
...
       hasContentLoading: false,
       contentError: false,
     });
-    const cred = wrapper.find('strong.credential');
-    expect(cred.text()).toContain('Cloud:');
-    done();
+
+    const chip = wrapper.find('CredentialChip');
+    expect(chip).toHaveLength(1);
+    expect(chip.prop('credential')).toEqual(
```
Comment (marshmalien) [inline awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.test.jsx:129]: I think this is another example of testing an implementation detail.

File: awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx [hunk 5]
```diff
@@ -1,35 +1,156 @@
...
+        <VariablesInput
+          css="margin: 20px 0"
+          id="job-variables"
+          readOnly
+          value={job.extra_vars}
+          rows={4}
+          label={i18n._(t`Variables`)}
+        />
+      )}
+      <VariablesInput
```
Comment (marshmalien) [inline awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx]: Do we always show the Artifacts text editor? I'm wondering if we should conditionally render it similar to how you are currently doing it for extra_vars.

File: awx/ui_next/src/screens/Job/Job.jsx [hunk 6]
```diff
@@ -106,19 +105,23 @@ class Job extends Component {
         <Card>
           {cardHeader}
           <Switch>
-            <Redirect from="/jobs/:id" to="/jobs/:id/details" exact />
-            {job && (
+            <Redirect
+              from="/jobs/:type/:id"
+              to="/jobs/:type/:id/details"
```
Comment (jakemcdermott) [inline awx/ui_next/src/screens/Job/Job.jsx:128]: We'll need to update the launch button component to make sure it continues to route users to the correct location: https://github.com/ansible/awx/blob/devel/awx/ui_next/src/components/LaunchButton/LaunchButton.jsx#L45

File: awx/ui_next/src/screens/Job/constants.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/* eslint-disable-next-line import/prefer-default-export */
+export const JOB_TYPE_URLS = {
```
Comment (jakemcdermott) [inline awx/ui_next/src/screens/Job/constants.js]: I learned the other day that there's actually a specific name for these - url path segment (or just url segment). 

How do you feel about `JOB_TYPE_URL_SEGMENTS` instead?

File: awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx [hunk 8]
```diff
@@ -206,13 +202,7 @@ class JobTemplateDetail extends Component {
...
                     <ChipGroup showOverflowAfter={5}>
                       {summary_fields.credentials.map(c => (
-                        <Chip key={c.id} isReadOnly>
-                          <strong className="credential">
-                            {c.kind ? credentialType(c.kind) : i18n._(t`Cloud`)}
-                            :
-                          </strong>
-                          {` ${c.name}`}
-                        </Chip>
+                        <CredentialChip key={c.id} credential={c} isReadOnly />
```
Comment (jakemcdermott) [inline awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx:205]: https://github.com/keithjgrant/awx/blob/222-job-results-details/awx/ui_next/src/components/Chip/CredentialChip.jsx#L11 is the tricky / unexpected part of showing this data (IMHO) and it looks like this new component is handling it so :+1:

**Review by jakemcdermott — CHANGES_REQUESTED** [review]
This is looking good! I left a comment about updating our launch component to make sure the redirects still work. Other than that, the rest of the things I mentioned are nitpicks / suggestions.

I think you'll need to rebase this onto latest devel to get the zuul tests passing.

File: awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx [hunk 9]
```diff
@@ -1,35 +1,164 @@
...
+  3: '3 (Debug)',
+  4: '4 (Connection Debug)',
+};
+
+function JobDetail({ job, i18n }) {
+  const {
+    job_template: jobTemplate,
+    project,
+    inventory,
+    instance_group: instanceGroup,
```
Comment (marshmalien) [inline awx/ui_next/src/screens/Job/JobDetail/JobDetail.jsx:37]: Nice. I really like this rename and destructure pattern.

File: awx/ui_next/src/screens/Job/Job.jsx (hunk 6 above)
Comment (keithjgrant) [reply awx/ui_next/src/screens/Job/Job.jsx:128]: I added some Routes with a redirect component that uses the unified jobs api to do a lookup and redirect to the canonical url of a job. So `/jobs/123/details` will now redirect to e.g. `/jobs/playbook/123/details`

**Review by jakemcdermott — APPROVED** [review]

**Pull request merged**